### PR TITLE
Add documentation on internal log message structure

### PR DESCRIPTION
**Pull request by rpkamp**
Addresses https://github.com/Seldaek/monolog/issues/516

### CODE REVIEW

File: doc/message-structure.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+Within monolog log messages are passed around as arrays, for example to processors or handlers.
+The table below describes which keys are always available for every log message.
+
+key        | type                      | description
+-----------|---------------------------|-------------------------------------------------------------------------------
+message    | string                    | The log message. This string may contain placeholders that will be replaced by variables from the context, e.g., "User %username% logged in" with `['username' => 'John']` as context will be written as "User John logged in".
+level      | int                       | Severity of the log message. See log levels described in [01-usage.md](01-usage.md).
+level_name | string                    | String representation of log level.
+context    | array                     | Arbirtrary data passed with the constuction of the message. For example the username of the current user or their IP addres.
```
Comment (stof) [inline doc/message-structure.md]: typo: ``arbitrary``

File: doc/message-structure.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+channel    | string                    | The channel this message was logged to.
+datetime   | Monolog\DateTimeImmutable | Date and time when the message was logged. Class extends `\DateTimeImmutable`.
+extra      | array                     | A placeholder array where processors can put additional data. Always available, but empty if there are no processors registered.
+
+Some handlers and/or processors may add additional keys to the array.
```
Comment (stof) [inline doc/message-structure.md]: Processors should not add new keys. This is not a supported usage pattern. They should store them in ``extra``.

And handlers are not adding keys to the record either (they could be turning the record into another data structure needed by the target system though of course)

**Comment by rpkamp** [comment]
Both good points, thank you!

I just fixed them.

How do you feel about lines 16 through 19? I couldn't find anywhere that this was the case, but it felt right intuitively. I mentioned it because I've been confused about the existence of two arrays that appear to serve the same purpose for quite some time.

**Comment by an1zhegorodov** [comment]
@rpkamp I recall I had a feeling like something is missing in Monolog documentation. That's it! Thanks!
And lines 16-19 make perfect sense. I'd be very surprised if that's not that case. May be @Seldaek can give us hint on this?

File: doc/message-structure.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+level      | int                       | Severity of the log message. See log levels described in [01-usage.md](01-usage.md).
+level_name | string                    | String representation of log level.
+context    | array                     | Arbitrary data passed with the constuction of the message. For example the username of the current user or their IP addres.
+channel    | string                    | The channel this message was logged to.
+datetime   | Monolog\DateTimeImmutable | Date and time when the message was logged. Class extends `\DateTimeImmutable`.
+extra      | array                     | A placeholder array where processors can put additional data. Always available, but empty if there are no processors registered.
+
+At first glance `context` and `extra` look very similar, and they are in the sense that they both carry arbitrary data that is related to log the log message somehow.
+The main difference is that `context` can be supplied in user land (it is the 3rd parameter to `Logger::addRecord()`) whereas `extra` is internal only and can be filled processors.
```
Comment (hkdobrev) [inline doc/message-structure.md]: I think the end of the sentence here should be something like "can be filled *by* processors`.

**Comment by rpkamp** [comment]
@hkdobrev nice catch! fixed it and fixed a few more typos while I was at it :)

**Comment by rpkamp** [comment]
ping @Seldaek

File: doc/message-structure.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# Log message structure
+
+Within monolog log messages are passed around as arrays, for example to processors or handlers.
+The table below describes which keys are always available for every log message.
+
+key        | type                      | description
+-----------|---------------------------|-------------------------------------------------------------------------------
+message    | string                    | The log message. This string may contain placeholders that will be replaced by variables from the context, e.g., "User %username% logged in" with `['username' => 'John']` as context will be written as "User John logged in".
```
Comment (Seldaek) [inline doc/message-structure.md]: This placeholder stuff is only true if you use the PsrLogMessageProcessor btw, maybe it should be mentioned.

File: doc/message-structure.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+level      | int                       | Severity of the log message. See log levels described in [01-usage.md](01-usage.md).
+level_name | string                    | String representation of log level.
+context    | array                     | Arbitrary data passed with the construction of the message. For example the username of the current user or their IP address.
+channel    | string                    | The channel this message was logged to.
```
Comment (Seldaek) [inline doc/message-structure.md]: Might be good to clarify this is equal to the channel name passed to `new Logger($channel)`?

**Comment by Seldaek** [comment]
Also would be cool to link to it from some of the other doc pages, otherwise looks good to me, thanks for the work :)

**Comment by Seldaek** [comment]
Thanks, I'll add a link to it in the README at least :)

**Pull request merged**